File: Optimal/Python/q189_rotateArray.py
class Solution:
    def rotate(self, nums: List[int], k: int) -> None:
        """
        Do not return anything, modify nums in-place instead.
        """
        l=len(nums)
        k = k%l

        nums[:] = nums[l-k:]+nums[:l-k]  #rotate array right by k places

#Beats 100% time wise and 98.27% memory wise

File: Optimal/Cpp/q189_rotateArray.cpp
//time wise - beats 100%

class Solution {
public:
    void rotate(vector<int>& nums, int k) {
        int l = nums.size();
        k = k % l;
        
        // Perform rotation in-place
        vector<int> temp(nums.begin() + l - k, nums.end());
        nums.erase(nums.begin() + l - k, nums.end());
        nums.insert(nums.begin(), temp.begin(), temp.end());
    }
};

//memory wise - beats 100%

class Solution {
public:
    void rotate(vector<int>& nums, int k) {
        int n=nums.size();
        k=k%n;
        reverse(nums.begin(),nums.end());
        reverse(nums.begin(),nums.begin()+k);
        reverse(nums.begin()+k,nums.end());
    }
};